lightning/src/ln/channelmanager.rs

```diff
@@ -3545,6 +3553,36 @@ where
 
 						true
 					});
+					if peer_state.ok_to_remove(true) {
+						pending_peers_awaiting_removal.push(counterparty_node_id);
+					}
+				}
+			}
+
+			// When a peer disconnects but still has channels, the peer's `peer_state` entry in the
+			// `per_peer_state` is not removed by the `peer_disconnected` function. If the channels
+			// of to that peer is later closed while still being disconnected (i.e. force closed),
+			// we therefore need to remove the peer from `peer_state` separately.
+			// To avoid having to take the `per_peer_state` `write` lock once the channels are
+			// closed, we instead remove such peers awaiting removal here on a timer, to limit the
+			// negative effects on parallelism as much as possible.
+			if pending_peers_awaiting_removal.len() > 0 {
+				let mut per_peer_state = self.per_peer_state.write().unwrap();
+				for counterparty_node_id in pending_peers_awaiting_removal {
+					match per_peer_state.entry(counterparty_node_id) {
+						hash_map::Entry::Occupied(entry) => {
+							// Remove the entry if the peer is still disconnected and we still
+							// have no channels to the peer.
+							let remove_entry = {
+								let peer_state = entry.get().lock().unwrap();
+								peer_state.ok_to_remove(true)
```

Comment:
So this should cover some of the memory-leak concern brought in https://github.com/lightningdevkit/rust-lightning/pull/1988#pullrequestreview-1283993335. However, I don't know if it's covering well the case where we have inbound/outbound disconnected peers with funded channels (post-`funding_transaction_generated_intern` and the insertion in `channel_by_id`, or `internal_funding_created` for inbound channel) not confirming after undermined time, and therefore occupying peer state space. Likely, they should be pruned after some separate timer more generous than every minute like `timer_tick_occured` . If correct, I think it can in another follow-up or its own PR, as it sounds separate and more work.